lib/SPIRV/SPIRVToOCL.h

```diff
@@ -0,0 +1,130 @@
+#include "llvm/PassSupport.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+
+#include <cstring>
+
+using namespace llvm;
+using namespace SPIRV;
+using namespace OCLUtil;
+
+namespace SPIRV {
+class SPIRVToOCL : public ModulePass, public InstVisitor<SPIRVToOCL> {
+protected:
+  SPIRVToOCL() : ModulePass(ID), M(nullptr), Ctx(nullptr) {
+  }
+public:
+  virtual bool runOnModule(Module &M) = 0;
+
+  void visitCallInst(CallInst &CI);
+
+  // SPIR-V reader should translate vector casts into OCL built-ins because
+  // such conversions are not defined neither by OpenCL C/C++ nor
+  // by SPIR 1.2/2.0 standards. So, it is safer to convert such casts into
+  // appropriate calls to conversion built-ins defined by the standards.
+  void visitCastInst(CastInst &CI);
+
+  /// Transform __spirv_ImageQuerySize[Lod] into vector of the same length
+  /// containing {[get_image_width | get_image_dim], get_image_array_size}
+  /// for all images except image1d_t which is always converted into
+  /// get_image_width returning scalar result.
+  void visitCallSPRIVImageQuerySize(CallInst *CI);
+
+  /// Transform __spirv_Group* to {work_group|sub_group}_*.
+  ///
+  /// Special handling of work_group_broadcast.
+  ///   __spirv_GroupBroadcast(a, vec3(x, y, z))
+  ///     =>
+  ///   work_group_broadcast(a, x, y, z)
+  ///
+  /// Transform OpenCL group builtin function names from group_
+  /// to workgroup_ and sub_group_.
+  /// Insert group operation part: reduce_/inclusive_scan_/exclusive_scan_
+  /// Transform the operation part:
+  ///    fadd/iadd/sadd => add
+  ///    fmax/smax => max
+  ///    fmin/smin => min
+  /// Keep umax/umin unchanged.
+  void visitCallSPIRVGroupBuiltin(CallInst *CI, Op OC);
+
+  /// Transform __spirv_{PipeOpName} to OCL pipe builtin functions.
+  void visitCallSPIRVPipeBuiltin(CallInst *CI, Op OC);
+
+  /// Transform __spirv_* builtins to OCL 2.0 builtins.
+  /// No change with arguments.
+  void visitCallSPIRVBuiltin(CallInst *CI, Op OC);
+
+  /// Translate mangled atomic type name: "atomic_" =>
+  ///   MangledAtomicTypeNamePrefix
+  void translateMangledAtomicTypeName();
+
+  /// Get prefix work_/sub_ for OCL group builtin functions.
+  /// Assuming the first argument of \param CI is a constant integer for
+  /// workgroup/subgroup scope enums.
+  std::string getGroupBuiltinPrefix(CallInst *CI);
+
+  /// Transform __spirv_Atomic* to atomic_*.
+  ///   __spirv_Atomic*(atomic_op, scope, sema, ops, ...) =>
+  ///      atomic_*(atomic_op, ops, ..., order(sema), map(scope))
+  virtual Instruction* visitCallSPIRVAtomicBuiltin(CallInst *CI, Op OC);
+
+  /// Transform __spirv_MemoryBarrier to:
+  /// - OCL2.0: atomic_work_item_fence.__spirv_MemoryBarrier(scope, sema) =>
+  ///       atomic_work_item_fence(flag(sema), order(sema), map(scope))
+  /// - OCL1.2: mem_fence
+  virtual void visitCallSPIRVMemoryBarrier(CallInst *CI) = 0;
+
+  static char ID;
+protected:
+  Module * M;
+  LLVMContext *Ctx;
+};
+}
```

Comment:
Github seems to think that there is a newline missing at the end of this file (and other files too), please check and fix.

Reply:
Done